lib/class-wp-theme-json.php

```diff
@@ -0,0 +1,953 @@
+	 * to be shared among all instances so we don't
+	 * process it twice.
+	 *
+	 * @var array
+	 */
+	private static $blocks_metadata = null;
+
+	/**
+	 * Data schema of each context within a theme.json.
+	 *
+	 * Example:
+	 *
+	 * {
+	 *   'context-one': {
+	 *     'styles': {
+	 *       'color': {
+	 *         'background': 'color'
+	 *       }
+	 *     },
+	 *     'settings': {
+	 *       'color': {
+	 *         'custom': true
+	 *       }
+	 *     }
+	 *   },
+	 *   'context-two': {
+	 *     'styles': {
+	 *       'color': {
+	 *         'link': 'color'
+	 *       }
+	 *     }
+	 *   }
+	 * }
+	 */
+	const SCHEMA = array(
+		'selector' => null,
+		'supports' => null,
+		'styles'   => array(
+			'color'      => array(
+				'background' => null,
+				'gradient'   => null,
+				'link'       => null,
+				'text'       => null,
+			),
+			'typography' => array(
+				'fontFamily'     => null,
+				'fontSize'       => null,
+				'fontStyle'      => null,
+				'fontWeight'     => null,
+				'lineHeight'     => null,
+				'textDecoration' => null,
+				'textTransform'  => null,
+			),
+		),
+		'settings' => array(
+			'color'      => array(
+				'custom'         => null,
+				'customGradient' => null,
+				'gradients'      => null,
+				'link'           => null,
+				'palette'        => null,
+			),
+			'spacing'    => array(
+				'customPadding' => null,
+				'units'         => null,
+			),
+			'typography' => array(
+				'customFontSize'   => null,
+				'customLineHeight' => null,
+				'dropCap'          => null,
+				'fontFamilies'     => null,
+				'fontSizes'        => null,
+				'fontStyles'       => null,
+				'fontWeights'      => null,
+				'textDecorations'  => null,
+				'textTransforms'   => null,
+			),
+			'custom'     => null,
+		),
+	);
+
+	/**
+	 * Presets are a set of values that serve
+	 * to bootstrap some styles: colors, font sizes, etc.
+	 *
+	 * They are a unkeyed array of values such as:
+	 *
+	 * ```php
+	 * array(
+	 *   array(
+	 *     'slug'      => 'unique-name-within-the-set',
+	 *     'name'      => 'Name for the UI',
+	 *     <value_key> => 'value'
+	 *   ),
+	 * )
+	 * ```
+	 *
+	 * This contains the necessary metadata to process them:
+	 *
+	 * - path          => where to find the preset in a theme.json context
+	 *
+	 * - value_key     => the key that represents the value
+	 *
+	 * - css_var_infix => infix to use in generating the CSS Custom Property. Example:
+	 *                   --wp--preset--<preset_infix>--<slug>: <preset_value>
+	 *
+	 * - classes      => array containing a structure with the classes to
+	 *                   generate for the presets. Each class should have
+	 *                   the class suffix and the property name. Example:
+	 *
+	 *                   .has-<slug>-<class_suffix> {
+	 *                       <property_name>: <preset_value>
+	 *                   }
+	 */
+	const PRESETS_METADATA = array(
+		array(
+			'path'          => array( 'settings', 'color', 'palette' ),
```

Comment:
I guess the presets are for sure always inside settings. Could we assume that and not include settings as part of the path?

Reply:
I prefer to be explicit so all data is together. Removing settings from here means that all pieces of code that process the presets need to have that piece of info.